docs/source/tutorial/extension.rst

```diff
@@ -0,0 +1,180 @@
+Create your own trainer extension
+=================================
+
+.. currentmodule:: chainer
+
+In this section, you will learn about the following things:
+
+- :ref:`basic`
+- How to create your own trainer extension
+
+    * :ref:`by defining a simple function<function>`
+    * :ref:`by defining a function decorated with @make_extension <decorator>`
+    * :ref:`by defining a class inherited from Extension class <class>`
+
+.. _basic:
+
+What is trainer Extension?
+--------------------------
+
+:class:`~chainer.training.Extension` is a callable object that takes a :class:`~chainer.training.Trainer` object as an argument. Adding an :class:`~chainer.training.Extension` to a :class:`~chainer.training.Trainer` using :meth:`~chainer.training.Trainer.extend` method, the :class:`~chainer.training.Extension` will be called at the given timing you specified by using ``trigger`` object (See the details in :ref:`trigger`)
+
+A :class:`~chainer.training.Trainer` object has all information used in a training loop, e.g., models, optimizers, updaters, iterators, and datasets, etc. So you can change the settings of optimizers
+
+
+.. _function:
+
+Write a simple function
+-----------------------
+
+You can make a new :class:`~chainer.training.Extension` by writing a simple function which takes :class:`~chainer.training.Trainer` object as its argument. For example, when you want to reduce the learning rate at specified timing during training, ``lr_drop`` extension can be written as follows:
+
+.. testcode::
+
+    def lr_drop(trainer):
+        trainer.updater.get_optimizer('main').lr *= 0.1
+
+Then you can add this function to a :class:`~chainer.training.Trainer` object via :meth:`~chainer.training.Trainer.extend` method.
+
+.. code-block:: python
+
+    trainer.extend(lr_drop, trigger=(10, 'epoch'))
+
+It lowers the learning rate every 10 epochs by multiplying 0.1 with the current learning rate.
+
+.. _decorator:
+
+Write a function decorated with @make_extension
+-----------------------------------------------
+
+:meth:`~chainer.training.make_extension` is a decorator that adds some attributes to a given function. For example, the simple extension we created above can be written in this form:
+
+.. testcode::
+
+    @make_extension(trigger=(10, 'epoch'))
+    def lr_drop(trainer):
+        trainer.updater.get_optimizer('main').lr *= 0.1
+
+The difference between the above one and this is whether it has a default ``trigger`` or not. In this latter case, :meth:`lr_drop` has its default ``trigger`` so that unless another ``trigger`` is specified via :meth:`~chainer.training.Trainer.extend` method, the ``trigger`` specified in :meth:`~chainer.training.make_extension` is used as default. So the code below acts the same as the former example, i.e., it reduces the learning rate every 10 epochs.
+
+.. code_block:: python
```

Comment:
Please fix this line.
https://github.com/chainer/chainer/pull/3649/files#diff-cbcaba2532e9f3a72ff2b0588ef31a40R39